node/src/state.rs

```diff
@@ -356,6 +368,71 @@ impl State {
         })
     }
 
+    pub fn prevalidate_block(
+        &self,
+        block: &ArcBlockWithHash,
+    ) -> Result<(), BlockPrevalidationError> {
+        let Some((genesis, cur_global_slot)) =
+            None.or_else(|| Some((self.genesis_block()?, self.cur_global_slot()?)))
+        else {
+            // we don't have genesis block. This should be impossible
+            // because we don't even know chain_id before we have genesis
+            // block, so we can't be connected to any peers from which
+            // we would receive a block.
+            return Err(BlockPrevalidationError::GenesisNotReady);
+        };
+
+        // received_at_valid_time
+        // https://github.com/minaprotocol/mina/blob/6af211ad58e9356f00ea4a636cea70aa8267c072/src/lib/consensus/proof_of_stake.ml#L2746
+        {
+            let block_global_slot = block.global_slot();
+
+            let delta = genesis.constants().delta.as_u32();
+            if cur_global_slot < block_global_slot {
+                // Too_early
+                return Err(BlockPrevalidationError::ReceivedTooEarly);
+            } else if cur_global_slot.saturating_sub(block_global_slot) > delta {
+                // Too_late
+                return Err(BlockPrevalidationError::ReceivedTooLate);
+            }
+        }
+
+        if block.header().genesis_state_hash() != genesis.hash() {
+            return Err(BlockPrevalidationError::InvalidGenesisProtocolState);
+        }
+
+        let (protocol_versions_are_valid, protocol_version_matches_daemon) = {
+            let min_transaction_version = 1.into();
+            let v = &block.header().current_protocol_version;
+            let nv = block
+                .header()
+                .proposed_protocol_version_opt
+                .as_ref()
+                .unwrap_or(v);
+
+            // Our version values are unsigned, so there is no need to check that the
+            // other parts are not negative.
+            let valid = v.transaction >= min_transaction_version
+                && nv.transaction >= min_transaction_version;
+            let compatible = v.transaction == PROTOCOL_VERSION.transaction
+                && v.network == PROTOCOL_VERSION.network;
+
+            (valid, compatible)
+        };
+
+        if !protocol_versions_are_valid {
+            return Err(BlockPrevalidationError::InvalidProtocolVersion);
+        } else if !protocol_version_matches_daemon {
+            return Err(BlockPrevalidationError::MismatchedProtocolVersion);
+        }
+
+        // TODO(tizoc): check for InvalidDeltaBlockChainProof
+        // https://github.com/MinaProtocol/mina/blob/d800da86a764d8d37ffb8964dd8d54d9f522b358/src/lib/mina_block/validation.ml#L369
+        // https://github.com/MinaProtocol/mina/blob/d800da86a764d8d37ffb8964dd8d54d9f522b358/src/lib/transition_chain_verifier/transition_chain_verifier.ml
+
```

Comment:
Can we bring back constants check here?. Ocaml node doesn't have it, because those constants are never actually used (ones from the block), but in our case, we do.

Reply:
@binier I am not sure if that will always be true (that these constants must match the genesis), but checking the current version of the OCaml node I don't see any logic that could change them.